[Final] Subscriber attributes Part 2: Backend

aboedo opened the pull request:
This is the second part of the subscriber attributes PR: 
It adds backend methods for posting to the subscriber attributes endpoint, as well as logic for sending in the unsynced attributes on the post receipt request. 

### Requirements: 
- [x] ~Part 1 #187~
- [x] ~Unit tests~

aboedo commented on Purchases/NSError+RCExtensions.m:
```diff
@@ -0,0 +1,37 @@
+
+@implementation NSError (RCExtensions)
+
+- (BOOL)shouldMarkSyncedKeyPresent {
```
I'm still not a fan of this name. first of all, this is true if and only if it the key is present **and** its value is true. I'm open to suggestions

aboedo commented on Purchases/Public/RCPurchases.m:
```diff
@@ -593,6 +593,7 @@ - (void)restoreTransactionsWithCompletionBlock:(nullable RCReceivePurchaserInfoB
                             discounts:nil
           presentedOfferingIdentifier:nil
                          observerMode:!self.finishTransactions
+                 subscriberAttributes:nil
```
part 4 (#190) updates this to send the real values

aboedo commented on PurchasesTests/BackendTests.swift:
```diff
@@ -92,12 +92,22 @@ class BackendTests: XCTestCase {
-                completion: { (purchaserInfo, error) in
-                    completionCalled = true
-                })
+        backend?.postReceiptData(receiptData,
```
these are mostly reformatting + adding the subscriber attributes param

aboedo replied:
tests specifically for subscriber attributes in `RCBackend` are in `BackendSubscriberAttributesTests`

aboedo commented on PurchasesTests/Mocks/MockHTTPClient.swift:
```diff
@@ -0,0 +1,38 @@
+import Purchases
+
+class MockHTTPClient: RCHTTPClient {
```
the mocks in these PRs are mostly autogenerated. It's nice because they also have very consistent formatting and naming, plus they're pretty complete.

aboedo commented on PurchasesTests/SubscriberAttributes/SubscriberAttributeTests.swift:
```diff
@@ -57,7 +57,6 @@ class SubscriberAttributeTests: XCTestCase {
 
         expect(receivedDictionary["key"] as? String) == key
         expect(receivedDictionary["value"] as? String) == value
-        expect(receivedDictionary["setTime"] as! Date) == now
```
I think this is an artifact from a rebase, I'll fix before the final PR is merged

aboedo replied:
https://github.com/RevenueCat/purchases-ios/pull/189/files#diff-9739dbeb09917a58be870452732ebfcaR60

vegaro commented on Purchases/Public/RCPurchasesErrorUtils.m:
```diff
@@ -146,6 +150,8 @@ static RCPurchasesErrorCode RCPurchasesErrorCodeFromRCBackendErrorCode(RCBackend
         case RCBackendUserIneligibleForPromoOffer:
             return RCIneligibleError;
+        case RCBackendInvalidSubscriberAttributes:
+            return RCBackendInvalidSubscriberAttributesError;
```
We normally try to map these backend errors to an error part of the `RCPurchasesErrorCode` enum, like in:
```
case RCBackendStoreProblem:
        case RCBackendPlayStoreQuotaExceeded:
        case RCBackendPlayStoreInvalidPackageName:
        case RCBackendPlayStoreGenericError:
            return RCStoreProblemError;
```
The reason behind that is that we expose these errors and we don't want the developers to have to handle all these backend errors separately. 

I think in this case we have to create one, which you already did 😄. We don't have one `RCPurchasesErrorCode` already that could group the backend errors when posting user attributes. I would change `RCBackendInvalidSubscriberAttributesError` to `RCInvalidSubscriberAttributesError` though to differentiate that error from the backend one, so that we can asign future `RCBackendErrorCode` related to attributes to that `RCPurchasesErrorCode.RCInvalidSubscriberAttributesError`.

aboedo replied:
sounds good, will do. thanks!

vegaro commented on Purchases/Public/RCPurchasesErrorUtils.m:
```diff
@@ -115,6 +117,8 @@
             return @"INVALID_APPLE_SUBSCRIPTION_KEY";
         case RCIneligibleError:
             return @"INVALID_OFFER";
```
Hmm, unrelated but I think this should be @"INELIGIBLE_ERROR", can you change it?

aboedo replied:
sure thing!

vegaro commented on Purchases/Public/RCPurchasesErrorUtils.m:
```diff
@@ -115,6 +117,8 @@
         case RCIneligibleError:
             return @"INVALID_OFFER";
+        case RCBackendInvalidSubscriberAttributesError:
+            return @"SUBSCRIBER_ATTRIBUTES_SYNC_ERROR";
```
I normally make these match the error code, so I would make it @"INVALID_SUBSCRIBER_ATTRIBUTES"

vegaro commented on Purchases/RCBackend.h:
```diff
@@ -87,6 +89,12 @@ presentedOfferingIdentifier:(nullable NSString *)offeringIdentifier
+                       appUserID:(NSString *)appUserID
+                      completion:(void (^)(NSError *))completion;
+
+extern NSErrorUserInfoKey const RCShouldMarkSyncedKey;
```
Should this go on top of the file? I am not sure what is the most common way of doing this 😄

aboedo replied:
I wasn't really sure what to go with here, happy to move it up 👍

vegaro commented on Purchases/RCBackend.m:
```diff
@@ -459,4 +467,66 @@ - (void)postOfferForSigning:(NSString *)offerIdentifier
+    if (statusCode > 300) {
+        BOOL isInternalServerError = statusCode >= 500;
+        NSDictionary *extraUserInfo = @{
+            RCShouldMarkSyncedKey: @(!isInternalServerError)
```
What do you think about naming as `RCShouldMarkAsSyncedKey`, or `RCSuccessfullySyncedKey`?

aboedo replied:
I like them. I'm going with `RCSuccessfullySyncedKey` because it better represents what happened, rather than what _should_ happen next

vegaro commented on Purchases/NSError+RCExtensions.m:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    return NO;
+}
```
I am a little bit confused about what's going on here. Can't you just check for  `RCShouldMarkSyncedKey` in the userInfo? For the network error there won't be any `RCShouldMarkSyncedKey`, so you can just do `if [self.userInfo[RCShouldMarkSyncedKey]) boolValue]`, since if it doesn't exist, it will evaluate to false. I would need to test that, but look into how it's done when posting the receipts RCPurchases:857.

I am also confused about the recursive call in line 27. I don't think the underlying errors will ever contain the should mark as synced key right?

aboedo replied:
I'll double-check the exact behavior, but the reason I added the recursive call in the first place was because the original error was getting wrapped inside another one, so a shallow search would bring a false negative

vegaro replied:
Was this fine then?

aboedo replied:
just re-tested, and you're right, this isn't necessary. The value in the user info comes in for the top-most error, not the underlying one. 
I think I might have gotten confused while testing for network errors ¯\_(ツ)_/¯
In any case, I killed the recursive call and simplified the checks

vegaro commented on PurchasesTests/SubscriberAttributes/BackendSubscriberAttributesTests.swift:
```diff
@@ -0,0 +1,231 @@
+                                         })
+        expect(self.mockHTTPClient.invokedPerformRequestCount) == 0
+    }
+
```
I would add some tests regarding the `RCShouldMarkSyncedKey` stuff. I think something like this should be enough:
1. On `503` it is false (you can add the assert to the existing test)
2. On  `304` it is true

aboedo replied:
good catch, added

aboedo commented:
@vegaro this is ready for another pass, updated from review comments. I changed the `updated_at` value to send milliseconds instead of seconds as well.

vegaro commented on PurchasesTests/SubscriberAttributes/BackendSubscriberAttributesTests.swift:
```diff
@@ -0,0 +1,293 @@
+//
+//        let receivedNSError = receivedError! as NSError
+//        expect(receivedNSError.code) == Purchases.ErrorCode.unknownBackendError.rawValue
+//    }
```
Why is this commented out? Is it supposed to work only in following PRs?

aboedo replied:
woooah, forgot about it, sorry!

aboedo replied:
it was meant for this one, forgot to uncomment

aboedo replied:
actually, it was a copy/paste while testing stuff, but it's duplicated from an existing test. killing

vegaro commented on Purchases/SubscriberAttributes/RCSubscriberAttribute.m:
```diff
@@ -79,7 +80,7 @@ - (instancetype)initWithDictionary:(NSDictionary <NSString *, NSObject *> *)dict
     return @{
         BACKEND_VALUE_KEY: self.value ?: @"",
-        BACKEND_TIMESTAMP_KEY: @(self.setTime.timeIntervalSince1970)
+        BACKEND_TIMESTAMP_KEY: @(self.setTime.millisecondsSince1970)
```
the value of the key has also changed to `updated_at_ms`

aboedo replied:
good catch, updated

vegaro commented on PurchasesTests/SubscriberAttributes/BackendSubscriberAttributesTests.swift:
```diff
@@ -57,11 +57,11 @@ class BackendSubscriberAttributesTests: XCTestCase {
             "attributes": [
                 subscriberAttribute1.key: [
-                    "updated_at": subscriberAttribute1.setTime.timeIntervalSince1970,
+                    "updated_at": (subscriberAttribute1.setTime as NSDate).millisecondsSince1970(),
```
Update to `updated_at_ms`

vegaro commented:
:shipit:
